### DISCUSSION TITLE
[@types/lodash] `union` has unusual behavior with `string` inputs

### DISCUSSION THREAD

**Discussion by zachhardesty7** [DefinitelyTyped/DefinitelyTyped#62881] [Issues with a @types package]
Hello! I ran across a weird behavior in Lodash's `union` function today and was wondering if the type is correct or if it needs changes. Let me elaborate.

The `union` function ([docs here](https://lodash.com/docs/4.17.15#union)) accepts any number of arguments of `ArrayLike` types ([see here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/b101e0a63930ce805fe5f93170233730ae37f81b/types/lodash/common/array.d.ts#L1564)). As I rediscovered today, `string` is an `ArrayLike` and allowed by TS to be passed to `union`. Here's [a quick playground](https://www.typescriptlang.org/play?#code/PTAECMFcEsBsBdTQHangTwA4FMmoCoDKAUCvNgE4BmAhgMa4CCFFN6AMtANbYA8+APlABvYqFAVsNACYB7ZLHShY2ZAHN4ACwBcoZJAC24SgG4xEqXIVKA2sl36jlALq78ZgL7Fi07HVg0kqBUkMh08NDyoKGRyPwCABTmIHiwKNjSoMysHNy4GDjmAHQlgTkAzrqi4uIq6loOhsYUZjWgdo1OFK6g7uYeNs7EAJRug2bEKojk5fAAjLqzFChqg6AAvO3OZjPzG9HIsQkARJrQx8MTU2jYswBMnc1rm+PEu3f7MfIJd5fe17sAMyPSjPLZvW7wQGfQ7fY67Y4AGnad2cyJsABZnMNiEA) using a simplified version of the declared type for `union` which only allows 1 parameter of type `ArrayLike`. This shows TS throwing an error for `number` but not `string`.

Now, while Lodash doesn't _error_ with `string` inputs, it certainly doesn't handle them as an `ArrayLike` and form a new array with unique chars. As you can see in [this REPL](https://runkit.com/zachhardesty/63581abb8f781a0008a5cc68), `string` inputs are simply ignored, not broken into their individual chars or anything else.
```ts
_.union("test", ..."test2", ["test3"]);
// expected? => ["t", "e", "s", "2", "test3"]
// actual => ["test3"]
```

Assuming Lodash is behaving correctly, the type for `union` (and presumably other functions) seems incorrect to me. We can very easily create valid Lodash code that TS does not allow, such as the last section of the TS Playground. Since `string` args are ignored, this should presumably be a valid assignment.
```ts
let test3: number[] = []
test3 = union("test", [2], [4])
// test3 === [2, 4]
```

Thoughts on this? Can it even be represented in TS? Does changing the type break too much code?

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
If I understand correctly, it sounds like we should disallow string
parameters. However I'm not sure if it's possible to allow
ArrayLike<string> but not string.

**Reply by zachhardesty7** [reply]
Your understanding matches what I believe to be the most intuitive behavior. Something like the following seems to work to prevent only exactly strings in my brief testing. There's probably a simpler version.
```ts
declare function union<T, T2 extends List<T> | null | undefined>(
  ...arrays: T2 extends string ? never : T2[]
): T[]
```

That being said, whether I like it or not, Lodash allows passing a `string` argument that it simply ignores. I don't know of any way to use generics that prevents TS from resolving the generic type to certain types (like `string`). To put it more simply, I don't know of a way to allow but ignore parameters of only a certain type.
